From dbbc704be2e16870e267781369f1eeb0b787623b Mon Sep 17 00:00:00 2001
From: Jan Nieuwenhuizen <janneke@gnu.org>
Date: Mon, 11 May 2009 16:02:14 +0200
Subject: [PATCH] OOXML: export autoFilter if it set, even if details are empty.  Fixes n#497559.

   * Modified     sc/source/filter/xlsx/excrecds.hxx
   * Modified     sc/source/filter/excel/excrecds.cxx
---
 sc/source/filter/xlsx/excrecds.hxx      |    1 +
 sc/source/filter/excel/excrecds.cxx |    9 +++++++--
 2 files changed, 8 insertions(+), 2 deletions(-)

diff --git sc/source/filter/inc/excrecds.hxx sc/source/filter/inc/excrecds.hxx
index 6ad0f71..57744dc 100644
--- sc/source/filter/inc/excrecds.hxx
+++ sc/source/filter/inc/excrecds.hxx
@@ -488,6 +488,7 @@ private:
     XclExpFiltermode*   pFilterMode;
     XclExpAutofilterinfo* pFilterInfo;
     ScRange                 maRef;
+    bool mbAutoFilter;
 };
 
 // ----------------------------------------------------------------------------
diff --git sc/source/filter/excel/excrecds.cxx sc/source/filter/excel/excrecds.cxx
index 6bf8e9e..171009a 100644
--- sc/source/filter/excel/excrecds.cxx
+++ sc/source/filter/excel/excrecds.cxx
@@ -883,6 +883,7 @@ ExcAutoFilterRecs::ExcAutoFilterRecs( const XclExpRoot& rRoot, SCTAB nTab ) :
     XclExpRoot( rRoot ),
     pFilterMode( NULL ),
     pFilterInfo( NULL )
+    , mbAutoFilter (false)
 {
     ScDBCollection& rDBColl = GetDatabaseRanges();
     XclExpNameManager& rNameMgr = GetNameManager();
@@ -983,6 +984,9 @@ ExcAutoFilterRecs::ExcAutoFilterRecs( const XclExpRoot& rRoot, SCTAB nTab ) :
             if( !maFilterList.IsEmpty() )
                 pFilterMode = new XclExpFiltermode;
             pFilterInfo = new XclExpAutofilterinfo( aRange.aStart, nColCnt );
+
+            if (maFilterList.IsEmpty () && !bConflict)
+                mbAutoFilter = true;
 		}
 	}
 }
@@ -1040,7 +1044,7 @@ void ExcAutoFilterRecs::Save( XclExpStream& rStrm )
 
 void ExcAutoFilterRecs::SaveXml( XclExpXmlStream& rStrm )
 {
-    if( maFilterList.IsEmpty() )
+    if( maFilterList.IsEmpty() && !mbAutoFilter )
         return;
 
     sax_fastparser::FSHelperPtr& rWorksheet = rStrm.GetCurrentStream();
@@ -1048,7 +1052,8 @@ void ExcAutoFilterRecs::SaveXml( XclExpXmlStream& rStrm )
             XML_ref,    XclXmlUtils::ToOString( maRef ).getStr(),
             FSEND );
     // OOXTODO: XML_extLst, XML_sortState
-    maFilterList.SaveXml( rStrm );
+    if( !maFilterList.IsEmpty() )
+        maFilterList.SaveXml( rStrm );
     rWorksheet->endElement( XML_autoFilter );
 }
 
-- 
1.6.0.rc1.49.g98a8

